[Ngo Xian Yi Justin] iP

whoisjustinngo opened the pull request:
<h1>Duke...</h1><h1>NEW AND IMPROVED 💯 💥 ❗ 👀 🔓 ✅ ✏️ 📝 </h1>

Feeling overwhelmed? Unproductive because your mind is constantly focusing on the next task and the task after that?
> "Anger is the ultimate destroyer of your own piece of mind" - Dalai Lama

and a cluttered mind is the ultimate destroyer of your productivity!

Try Duke! All you have to do is:
1. download the .jar file from [here](https://github.com/whoisjustinngo/ip/releases/tag/v0.1)
2. double click on "iP.jar" or, in terminal, type:
`cd [name of folder containing iP.jar file]` 
hit enter, then type
`java -jar iP.jar`
and hit enter
3.  start tracking those tasks!

Duke:
* provides support for three types of tasks: simple to-dos, events, and deadlines
* has a text-based UI
* is easy to pickup
* is intuitive
* requires absolutely no installation or account creation
* can work offline

**For the low price of $0** you can have access to **DUKE**, the only virtual task manager you need!

Some features of Duke include:
- [x] add and delete tasks 📝 
- [x] mark task as completed ✅ 
- [x] list all tasks currently in list 🔢 
- [x] search for tasks based on keywords 🔍 
- [x] save tasks to local data for booting up in future runs 😮 
- [ ] sorting by event date or deadline (coming soon 👀 )
- [ ] calendar app integration (coming soon 👀 )

If you are the type of person that understands code better than plain text...:
```java
public static getBestTaskManager(Person you) throws JustTryDukeException {
   if (you.isUnproductive) {
        if (you.require(goodTaskManager)) {
            if (you.isBroke) {
                return "Duke";
           }
        }
    }
    throw new JustTryDukeException();
}
```
This program will help you keep track of your tasks, so you can focus on what really matters... completing those tasks!

Commit:
Add Gradle support

whoisjustinngo committed:
Duke.java: Implement skeletal version (initial commit).

whoisjustinngo committed:
MyList.java: init custom list class

whoisjustinngo committed:
MyList.java: add return statement to indicate item added

whoisjustinngo committed:
Duke.java:  implement level 2 - add, list functions

whoisjustinngo committed:
Init new Item class and modify MyList to use it

whoisjustinngo committed:
add level 3: ability to mark tasks as done

whoisjustinngo committed:
Initialise Deadline, Event, ToDo classes, change Task to abstract class

whoisjustinngo committed:
Implement level 4 - todo, event, deadline

whoisjustinngo committed:
MyList.java: add feature to prevent duplicate tasks

whoisjustinngo committed:
Set up input and expected file, configure script for auto testing

whoisjustinngo committed:
Add checking if task is already completed

whoisjustinngo committed:
Initialise exceptions for Duke

whoisjustinngo committed:
Modify program to handle invalid input using exceptions

whoisjustinngo committed:
Add delete task function

whoisjustinngo committed:
Duke.java: removed irrelevant comment

whoisjustinngo committed:
Change code to suit coding convention and for consistency

whoisjustinngo committed:
Implement level 7 functionality (saving to and loading from disk)

whoisjustinngo committed:
Implement level 8: handling date and times using LocalDateTime, LocalDate, LocalTime classes

whoisjustinngo committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Implement level 7 functionality (saving to and loading from disk)

whoisjustinngo committed:
Merge branch 'branch-Level-8'; deconflict merge conflicts

* branch-Level-8:
  Implement level 8: handling date and times using LocalDateTime, LocalDate, LocalTime classes

# Conflicts (resolved)
#	src/main/java/Duke.java
#	src/main/java/MyList.java

whoisjustinngo committed:
Refactor the code to extract out closely related code as classes

whoisjustinngo committed:
Organise the classes into java packages

Package task contains class Task and its subclasses
Package exception contains custom DukeException class that contains all custom exceptions
Package utility contains the custom classes that enable the main program to read, parse, organise and return data

whoisjustinngo committed:
Implement junit tests for Storage.java and Parser.java

whoisjustinngo committed:
Change name of log file

whoisjustinngo committed:
Edit code to follow coding convention and make generally more readable

whoisjustinngo committed:
Add feature to search for tasks using input keywords

whoisjustinngo committed:
Add javadocs

whoisjustinngo committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Edit code to follow coding convention and make generally more readable

whoisjustinngo committed:
Merge branch 'branch-Level-9'; resolve conflicts

* branch-Level-9:
  Add feature to search for tasks using input keywords

# Conflicts:
#	src/main/java/duke/utility/TaskList.java

whoisjustinngo committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add javadocs

# Conflicts:
#	src/main/java/duke/exception/DukeException.java
#	src/main/java/duke/utility/Parser.java
#	src/main/java/duke/utility/Storage.java

whoisjustinngo committed:
Parser.java: refactor parseNewTask method to make it less convoluted
TaskList.java: rename methods to follow code convention

whoisjustinngo committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  Add Gradle support

whoisjustinngo committed:
Enable running and building of duke using gradle (minimal for now)

whoisjustinngo committed:
Add support for checkstyle plugin; clean up code to meet coding standards

whoisjustinngo committed:
Merge branch 'A-CheckStyle'

* A-CheckStyle:
  Add support for checkstyle plugin; clean up code to meet coding standards

TLChicken commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,75 @@
+
+    /**
+     * An exception that is thrown if the user specifies an invalid task number to delete, e.g. if the task number
+     * specified is {@literal <} 0 or {@literal >} the number of tasks.
```
Excellent JavaDocs! Great job on making good use of all the available tags! 👍

TLChicken commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,54 @@
+     * Gets the completion status of the task.
+     * @return the completion status of the task.
+     */
+    public boolean getIsCompleted() {
```
Maybe a isTaskCompleted could be a better name for this method? 

Reference: Boolean variables/methods should be named to sound like booleans

Samuel-bit-prog replied:
Maybe change isCompleted to protected instead of private. Can remove this method.

TLChicken commented on src/main/java/duke/utility/Ui.java:
```diff
@@ -0,0 +1,43 @@
+        String welcomeMessage = "Hello I'm Duke!\nWhat can I do for you?";
+        this.printMessage(welcomeMessage);
+
+        boolean shouldContinue = true;
```
Maybe this boolean could be made more specific by naming it 'shouldContinueListening' instead of shouldContinue, it may be easier for readers to know what is being continued.

TLChicken commented on src/main/java/duke/utility/Storage.java:
```diff
@@ -0,0 +1,126 @@
+        sc.close();
+    }
+
+    void deleteTaskLogEntry(int lineIdx) throws IOException {
```
Perhaps the variable lineIdx could be explained in some Javadocs or maybe the variable name could be more descriptive? It may not be obvious what Idx means to readers.

TLChicken commented on src/main/java/duke/utility/TaskList.java:
```diff
@@ -0,0 +1,131 @@
+public class TaskList {
+
+    protected final HashSet<String> existingTasks;
+    private final List<Task> tasks;
```
Maybe the variable name tasks could be made more specific to describe what is being stored in the List, because having it might be unclear what this variable is being used for especially since it is placed under another variable with a similar name.

TLChicken reviewed (commented):
LGTM! Overall the JavaDocs were excellent! The naming of your variables are also mostly consistent except for a few of them, some of which I pointed out.

whoisjustinngo committed:
Add a GUI to Duke

whoisjustinngo committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add a GUI to Duke

Samuel-bit-prog commented on src/main/java/duke/utility/Parser.java:
```diff
@@ -0,0 +1,145 @@
+            default:
+                return this.parseNewTask(command.trim()); // default is add new duke.task
+            }
+        } catch (DukeException ex) {
```
Perhaps have the other exceptions put into the DukeException class and use enum

Samuel-bit-prog commented on src/main/java/duke/utility/Storage.java:
```diff
@@ -0,0 +1,128 @@
+        return prevTasks;
+    }
+
+    void append(String e, String f, String taskName, String dateTime) throws IOException {
```
perhaps add some javadocs for the methods below like append and changeTaskLogToCompleted

Samuel-bit-prog commented on src/main/java/duke/utility/TaskList.java:
```diff
@@ -0,0 +1,131 @@
+        return sb.toString();
+    }
+
+    protected int getTaskIndex(String taskName) throws DukeException.NoSuchTaskException {
```
Add javadocs for the methods below getTaskIndex and search

Samuel-bit-prog reviewed (commented):
LGTM

whoisjustinngo committed:
Fix bugs related to storage and parsing saved log file

whoisjustinngo committed:
Make saving and loading of tasks OS-independent

whoisjustinngo committed:
Use assert feature to document important assumptions that should hold at
various points

There are some parts in the code that require specific restricted input
from the user. For example, when entering the line number of the task to
delete, the number needs to be a valid number i.e. greater than 0 but
less than or equal to the total number of tasks.

Not validating user input could break code execution or lead to
unexpected outcomes.

Let's add assert statements at the various junctures that require data
validation to catch any invalid user input.

This will help indicate to us the various precautions to take in
anticipation of invalid input from the user.

As far as I can tell, all the potential areas where input validation
could be an issue are either dealt with exceptions or assert statements.

whoisjustinngo committed:
Refactor code to improve code quality

The methods in some classes (specifically those in the utility package)
have the same functionality. Some methods that deal with processing
user input also has convoluted if-else cases. Some input validation was
also done at the wrong place, e.g. Parser checking for duplicate tasks
when TaskList is the one in-charge of adding the task. The inclusion
of many messages in the main body is distracting and causes the main
code body to be cluttered.

The common functionalities across methods leads to code duplication.
Extensive if-else clauses and long methods inhibit readability and
understandability. Having some functionalities in the wrong class might
violate the tell-dont-ask principle.

Let's refactor the code so that everything is in its right place and
clean up the presentation.
* extract duplicate functionalities as a common method
* simplify if-else clauses and use guard clauses wherever necessary so
that code is easier to follow
* shorten long methods by extracting out auxiliary functionalities to
helper methods
* extract messages as constants which are placed away from the main body
of code
* ensure that the tell-dont-ask principle is followed by delegating the
job to the right class (e.g. validation for duplicate tasks should be
done by TaskList)
* other general code quality improvements

Improving code quality and readability makes debugging and further
developing much easier and smooth.

whoisjustinngo committed:
Merge pull request #1 from whoisjustinngo/branch-A-Assertions

Use assert feature for input validation

whoisjustinngo committed:
Merge branch 'master' into branch-A-CodeQuality

whoisjustinngo committed:
Merge branch 'master' of https://github.com/whoisjustinngo/ip

* 'master' of https://github.com/whoisjustinngo/ip:
  Use assert feature to document important assumptions that should hold at various points

whoisjustinngo committed:
Merge commit 'cdea20daad7ad3cdec29e61f4119f2eed065ada4'

* commit 'cdea20daad7ad3cdec29e61f4119f2eed065ada4':
  Refactor code to improve code quality

whoisjustinngo committed:
Merge branch 'branch-A-CodeQuality' of https://github.com/whoisjustinngo/ip into branch-A-CodeQuality

* 'branch-A-CodeQuality' of https://github.com/whoisjustinngo/ip:
  Use assert feature to document important assumptions that should hold at various points

whoisjustinngo committed:
Merge branch 'branch-A-CodeQuality'

* branch-A-CodeQuality:

whoisjustinngo committed:
Add new function to view all tasks (events/deadlines) occurring on a
specified date

whoisjustinngo committed:
Merge branch 'branch-B-ViewSchedules'

* branch-B-ViewSchedules:
  Add new function to view all tasks (events/deadlines) occurring on a specified date

whoisjustinngo committed:
Set theme jekyll-theme-tactile

whoisjustinngo committed:
Update README.md and add Ui.png for webpage

whoisjustinngo committed:
Merge branch 'master' of https://github.com/whoisjustinngo/ip

* 'master' of https://github.com/whoisjustinngo/ip:
  Set theme jekyll-theme-tactile

whoisjustinngo committed:
Set theme jekyll-theme-minimal

whoisjustinngo committed:
Merge branch 'master' of https://github.com/whoisjustinngo/ip

* 'master' of https://github.com/whoisjustinngo/ip:
  Set theme jekyll-theme-minimal

whoisjustinngo committed:
Improve Ui and user experience

whoisjustinngo committed:
Fix typo in README.md (for the user guide)

whoisjustinngo committed:
Update README.md to highlight the issue regarding disk access
permissions when running using double click

whoisjustinngo committed:
Fix formatting issues

whoisjustinngo committed:
Make minor changes to clean up code

whoisjustinngo committed:
Update README.md
